### PR TITLE
feat(async_flow): add an asynchronous streaming RL recipe

### PR DESCRIPTION
**Pull request by hongti**
## Background & Motivation

verl's existing RL pipeline runs synchronously: Rollout, Reference, Reward, and Trainer execute serially, leaving cluster resources idle across stages. In long-sequence GRPO scenarios (e.g. `prompt:2k → response:16k`), NPU utilization is significantly bottlenecked by this synchronous pipeline.

This PR introduces **AsyncFlow** under `recipe/async_flow/`, a training-inference asynchronous pipeline that decouples Rollout from downstream workers, combined with staleness tolerance and asynchronous weight synchronization, delivering substantial throughput gains for RLHF/GRPO.

## Key Changes

New directory `recipe/async_flow/` (52 files, ~13k LOC), built around five core components:

- **AgentLoop Layer** (`agent_loop/`): `AsyncFlowAgentLoopManager` + `AgentLoopWorker`, an asynchronous generation layer with `model_version` tracking.
- **TransferQueue (TQ_AsyncFlow)** (`utils/transfer_queue/`): A high-throughput data hub purpose-built for AsyncFlow. Control-plane / data-plane separation with distributed storage; introduces `topic` abstraction, reference-counted deduplication, dynamic capacity scaling, columnar metadata, end-to-end zero-copy, and multi-shard parallel transfer.
- **Four independent Workers** (`workers/`):
    
    - `ActorForwardWorker`: `old_log_probs` / `entropys`
    - `ReferenceForwardWorker`: `reference_logprobs`
    - `RewardAdvWorker`: `reward` / `advantage` (GRPO normalization)
    - `ActorTrainWorker`: PPO loss + parameter update
    
    Each worker runs an asyncio event loop in a dedicated daemon thread via `AsyncWorkerMixin`; `DataDispatchStrategy` abstracts FSDP / NONE and other parallel backends through the strategy pattern.
- **CheckpointEngine** (`utils/checkpoint_engine/`): HCCL-based asynchronous weight broadcast. Tensor metadata is transferred via ZMQ PUB/SUB; a double-buffered send/recv pipeline overlaps communication with hand-off; bucketed transfer (`update_weights_bucket_megabytes`) batches small tensors.
- **FlowControlQueue**: Staleness-control queue with `max_capacity = (staleness + 1) × batch_size`, capping in-flight requests. `staleness=0` degenerates to strict on-policy.

Entrypoints / configuration:

- `grpo_main.py`, `async_flow_trainer.py`, `config.py`, `config/async_grpo.yaml`: Reuse verl's Hydra config system, adding `async_resources.*` and other async-specific knobs.
- `vllm_rollout/vllm_async_server.py`: Async vLLM rollout integration.
- Launch scripts: `run_async_grpo_qwen{0.5b,7b}_gsm8k_{local,npu}.sh`, `run_qwen2_5_7b_grpo_npu.sh`.
- Observability: `utils/metric/prometheus*.py`, driven by `PROMETHEUS_METRICS_*` env vars.
- Docs: `README.md`, `docs/architecture.md`, `docs/performance.md`.

## Performance

GRPO with `prompt:2k → response:16k`:

| Metric | verl On-Policy | AsyncFlow | Fully Async |
| --- | --- | --- | --- |
| NPU partition | 64 | 52-2-2-8 | 48-16 |
| perf/throughput | 59.3 | **226.8** | 149.66 |
| Speedup | / | **3.81×** | 2.52× |

On accuracy, async RL with `staleness ≤ 2` matches or slightly exceeds synchronous RL (see the performance doc).

## Compatibility

- All changes are isolated under `recipe/async_flow/` and do not touch verl's main pipeline.
- Dependencies are listed in `REQUIRED_VERL.txt` . The current CheckpointEngine implementation requires Ascend NPU + HCCL.
- Reuses verl's Hydra config system; existing training scripts are unaffected.

## Acknowledgments
Huawei Advanced Computing and Storage Laboratory, Huawei Large Model System Engineering Lab, Huawei Technology Development Dept

-------------
Co-authored-by: nuerxiati <738457498@qq.com>
Co-authored-by: Li-Yongwen <1310439159@qq.com>
Co-authored-by: beirong8kmiles <zhoubeirong@huawei.com>
Co-authored-by: wucong25 <412916467@qq.com>
Co-authored-by: MatrixPlayer <wujianping.2006@163.com>

### CODE REVIEW

File: async_flow/agent_loop/__init__.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#
```
Comment (ji-huazhong) [inline async_flow/agent_loop/__init__.py:1]: Please rename the directory `async_flow` to `asyncflow`.

Comment (hongti) [reply async_flow/agent_loop/__init__.py:1]: I noticed other dir in verl-project are using `snake_case` style. So why should we change the naming style here?

**Review by wuxibin89 — APPROVED** [review]
